query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java

```diff
@@ -120,11 +129,11 @@ boolean hasIndex(final Class<?> c) {
       return searchFactory.getIndexBindings().get(c) != null;
    }
 
-   private void handleClusterRegistryRegistration(final Class<?> clazz) {
+   private CompletionStage<Void> handleClusterRegistryRegistration(final Class<?> clazz) {
       if (hasIndex(clazz)) {
-         return;
+         return null;
```

Comment:
return a completion stage that contains null value instead of null ?

Reply:
The listener code allows for a null return which means it is already complete. It is a more performant way than returning a completed `CompletionStage`.

Reply:
Is it really better than `completedNull()`? The return value is never used :)

Reply:
The return value is indeed used in CacheNotifierImpl as null is a valid value for listeners as documented.

Reply:
It wasn't when I looked :)

   @Listener
   final class CacheListener {

      @CacheEntryCreated
      public void created(CacheEntryCreatedEvent<KeyValuePair<String, Class>, Boolean> e) {
         if (!e.isOriginLocal() && !e.isPre() && e.getValue()) {
            handleClusterRegistryRegistration(e.getKey().getValue());
         }
      }

      @CacheEntryModified
      public void modified(CacheEntryModifiedEvent<KeyValuePair<String, Class>, Boolean> e) {
         if (!e.isOriginLocal() && !e.isPre() && e.getValue()) {
            handleClusterRegistryRegistration(e.getKey().getValue());
         }
      }
   }

Reply:
I clarified on the PR directly.